### ISSUE TITLE
Remove upper bound on Python version constraint

### ISSUE THREAD

**Issue by TomasBeuzen** [executablebooks/mdit-py-plugins#36]
### Describe the problem/need and solution

Firstly, as always, thanks for all the time and effort into supporting this project!

Upper version constraints are generally dissuaded unless there's a strong reason for using them (e.g., [PyPA PR](https://github.com/pypa/packaging.python.org/pull/850), Henry Schreiner's [excellent blog post](https://iscinumpy.dev/post/bound-version-constraints/)). Having an upper bound can make it difficult to use this package as a dependency in packages that do no have the same constraint. e.g., if I try to add it to a `poetry` project with Python >= 3.9:

```
The current project's Python requirement (>=3.9) is not compatible with some of the required packages Python requirement:
    - mdit-py-plugins requires Python ~=3.6, so it will not be satisfied for Python >=4.0
```

I note that [MyST-NB](https://github.com/executablebooks/MyST-NB/blob/5f4b8f6f5220caef0cf9f8c64294a2a831d8983e/setup.cfg#L53) (`python_requires = >=3.6`) and [MyST-Parser](https://github.com/executablebooks/MyST-Parser/blob/7654462dc8521c0090478efa4dcfba6227e97a84/setup.cfg#L47) (python_requires = >=3.7) and a current PR for [markdown-it-py](https://github.com/executablebooks/markdown-it-py/pull/194/files#r792053720) (python_requires = >=3.7) all do not have an upper version constraint. Unless there's a reason the Python version constraint is currently implemented with an upper bound, it would be great if we could remove it.

### Guide for implementation

Change [python_requires = ~=3.6](https://github.com/executablebooks/mdit-py-plugins/blob/aba72e6347e0c7ae8df9111a7d69cf0ab4bbf961/setup.cfg#L33) in `mdit-py-plugins/setup.cfg` to:

```
python_requires = >=3.6
```

More than happy to make this simple PR if it would be helpful/you agree :)

### Tasks and updates

- [ ] Remove upper version constraint on Python

**Comment by cjolowicz** [comment]
Just submitted #40 to address this.

I'd love to see this fixed. The current behavior forces any project adopting MyST to add a Python version cap, a discouraged and problematic practice.

cc @chrisjsewell @choldgraf 

Thanks!

**Comment by chrisjsewell** [comment]
Its not the end of the world to do (I doubt python 4 is coming out any time soon lol), but honestly I consider this a bug of poetry

**Comment by cjolowicz** [comment]
> Currently, this package forces any Poetry-managed project adopting MyST to add a Python version cap.

I have to correct myself here: Poetry projects can add myst-parser with an environment marker for the Python version to avoid the resolution failure:

```
poetry add --dev --python='<4' myst-parser
```

> Its not the end of the world to do (I doubt python 4 is coming out any time soon lol), but honestly I consider this a bug of poetry

Poetry is not at fault here. If this package says it won't work on Python 4 and later, should projects supporting all Python versions be able to depend on it? Poetry says no *unless* your environment has a Python version <4. Poetry may be strict here, but I would not consider its behavior a bug.

**Comment by hukkin** [comment]
I agree with @cjolowicz that Poetry is in the right and it is not a bug. Poetry is very strict about dependency resolution, but also correct. If a dependency is declared to only work with Python <4, then that same restriction logically transitions to any project using that dependency.

It should be noted that when looking at Python versioning from SemVer perspective, a Python major version is actually **not** SemVer major version. Or well, it is, but so is a minor version. I.e. moving from version 3.`x` to 3.`x+1` contains no guarantees of backwards compatibility.

If we really want to ensure that our application always works in the future, we should be pinning `python~=3.6.0` (which I'm not asking for :smile: ). But that brings me to another thing which is that `mdit-py-plugins` is not an application but a library. IMO libraries should never have upper pins, applications should be responsible for pinning and breakage of their users.

**Comment by chrisjsewell** [comment]
> MO libraries should never have upper pins (unless there is a known incompatibility). Applications should be responsible for pinning and breakage of their users.

Yes I've read the [blog post](https://iscinumpy.dev/post/bound-version-constraints/) @hukkin 😉, the same blog post the says basically never to use poetry.
Joking aside, I've already said we can remove the python pin, so lets not go down that rabbit hole.
Plus, we are on a trajectory to remove `attrs` from markdown-it-py, so then that will have no dependencies, which is basically the best way to not have any incompatibilities lol

**Comment by hukkin** [comment]
Haha sorry, TL;DR myself :smile:

**Comment by choldgraf** [comment]
FWIW, I'm also +1 on removing the cap on Python. We should certainly document something like "We make no promises about this tool's functionality for a version of Python that is less than N months old".

By the way - we are discussing this issue more broadly in https://github.com/executablebooks/meta/issues/648 in case others want to provide their thoughts / experiences / etc

**Comment by agoose77** [comment]
Just to add some notes for posterity here - the "broken" part of Poetry is the fact that it adds an upper-bound to `python_version` even if the author doesn't know that their package won't work on 4.0 

Python doesn't follow semver AFAICT, and makes no promises that 3.X→4.0 will be worse than 3.X-1→3.X. If every Poetry package started imposing arbitrary 3.X constraints, we'd be seeing this problem a lot more frequently.

The fact that the dependency solvers also start backtracking whenever they encounter a `python_version` is less clearly broken; logically it's doing the "right" thing, but our intentions don't always align with this. I think Henry's general message is that 
- compatibility tends to be discovered: I know version X breaks my package, so that's forbidden
- changes tend to be step functions: this feature was only added in verison X, so I need a lower bound

In this model, Python solvers would generally benefit from developers not upper capping *any* third-party dependencies, and using a forward-only solver. That's a big sell, though, so we'll probably just muddle through in the mean time. :laughing:

**Comment by chrisjsewell** [comment]
> Python solvers would generally benefit from developers not upper capping any third-party dependencies

thats fine .... if the maintainers are willing to immediately fix their packages whenever a third-party dependency release inevitably breaks its.
I'm not

> a forward-only solver

How is this even a solver? you would simply take the latest release of every package and that would "solve" it 🤷

**Comment by agoose77** [comment]
> hats fine .... if the maintainers are willing to immediately fix their packages whenever a third-party dependency release inevitably breaks its. I'm not

It's a lose-lose unfortunately; either you cap everything, and anyone with vaguely deep dependency trees gets stuck on old packages (best case), or hits dependency conflicts (worst case) and can't solve.

In my view, releasing upper caps is better; we can give users the tools and documentation to apply constraints (we already have the tools, `pip` has constraint files). Within your own packages, where you explicitly choose release cadence and compatabilty schedules, the argument is less important.

> How is this even a solver? you would simply take the latest release of every package and that would "solve" it shrug

By forward-only, I mean no backtracking to find more permissive constraints. What happens frequently now is that older packages have "worse" metadata (i.e. no upper bounds), and solvers end up finding it and installing a 5 year old package instead of raising an error to point out a solve issue. This is most visible with the Python version, but also applies to general dependencies. The solution is to ask solvers never to choose an older version of a package with higher upper bounds - this is nearly always not intentional.

**Comment by chrisjsewell** [comment]
> In my view, releasing upper caps is better; we can give users the tools and documentation to apply constraints

Expert users yeh, not casual user, of the kind the mostly use this package.
Most people do not care about this at all, they just want a **stable** that they can install and it works. Its only the "vocal minority" of advanced users, that want to be at the bleeding edge, that ever complains about this 🤷  

You remove upper-pinnings, and your package becomes inherently unstable; people can install it one day, and it works fine, the next some changes something in a dependency and breaks everything.

**Comment by agoose77** [comment]
I'm not sure that's true. Casual users are just as likely to hit solve errors, and they will raise a ticket the moment they encounter one. This discussion isn't about bleeding edge - it's just about a ticking time bomb of dependencies. Most of us have already seen this in the last couple of years with `jedi` as one high-profile example. It would be worse, but there is a general consensus of not upper capping in the big projects.

If your upper bounds are too restrictive, then you increase your own maintenance burden by needing to release more often just to bump caps. Or, you don't do that, and then users are more likely to install someone else's package with incompatible pins. A simple HEP environment with JupyterLab + some numerical packages pulls in 90+ packages. Your probability of an unfixable solve error increases rapidly.

The proper solution to this is for developers to be as open as possible about when they'll break parts of the API - more explicit than "something might break in the next version", which is overly restrictive. That's basically what deprecation warnings are. The best packages would start providing future APIs alongside the old ones, and let dependents soft-transition. With this kind of comprehensive discussion, we can make more informed decisions about whether/when to cap things, and what kinds of caps to use. Semver is a massive shovel in an ecosystem where we have no solutions if it goes wrong. (Well, actually `pdm` lets you brute force a solution but it's a highly unpleasant experience as first the solver must be exhausted).

**Comment by chrisjsewell** [comment]
> I'm not sure that's true.

I am

> Most of us have already seen this in the last couple of years with jedi as one high-profile example. 

you're cherry-picking, there was just as big an issue with sphinx-builds when jinja broke everyone's book builds due to having no upper cap: https://github.com/pallets/jinja/issues?q=is%3Aissue+sort%3Aupdated-desc+markup+is%3Aclosed

>  the Jupyter APIs let us use external kernel-specs, so the Python environment can sit aside from the MyST tooling.

yes this anwser I've already said multiple times, and has nothing to do with pinning

If jupyter-book didn't pin myst-parser, it would have been broken since the release of 0.14.0 on 04 May 2021, and its still yet to be updated.

If you are willing to be responsible for fixing everything across EBP, as soon as there is an issue, then great we can unpin everything
